internal/provider/common.go

```diff
@@ -9,30 +10,48 @@ import (
 	"github.com/containerd/containerd/platforms"
 	"github.com/crazy-max/diun/v4/internal/model"
 	"github.com/crazy-max/diun/v4/pkg/registry"
-	"github.com/pkg/errors"
 )
 
 var (
 	metadataKeyChars  = `a-zA-Z0-9_`
 	metadataKeyRegexp = regexp.MustCompile(`^[` + metadataKeyChars + `]+$`)
+	ErrInvalidLabel   = errors.New("invalid label error")
 )
 
+func UpdateMap(m1, m2 map[string]string) map[string]string {
+	if m1 == nil && m2 != nil {
+		return m2
+	}
+
+	if m1 != nil && m2 == nil {
+		return m1
+	}
+
+	if m1 == nil && m2 == nil {
+		return nil
+	}
+
+	for key, value := range m2 {
+		m1[key] = value
+	}
+
+	return m1
+}
+
 // ValidateImage returns a standard image through Docker labels
-func ValidateImage(image string, metadata, labels map[string]string, watchByDef bool) (img model.Image, err error) {
+func ValidateImage(image string, metadata, labels map[string]string, watchByDef bool, imageDefaults model.Image) (img model.Image, err error) {
 	if i := strings.Index(image, "@sha256:"); i > 0 {
 		image = image[:i]
 	}
-	img = model.Image{
-		Name:     image,
-		Metadata: metadata,
-		NotifyOn: model.NotifyOnDefaults,
-		SortTags: registry.SortTagReverse,
-	}
+
+	img = imageDefaults
+	img.Name = image
+	img.Metadata = nil
 
 	if enableStr, ok := labels["diun.enable"]; ok {
 		enable, err := strconv.ParseBool(enableStr)
 		if err != nil {
-			return img, fmt.Errorf("cannot parse %s value of label diun.enable", enableStr)
+			return img, fmt.Errorf("cannot parse %q value of label diun.enable: %w", enableStr, ErrInvalidLabel)
```

Comment:
This is so that any errors due to a label return an error derived from the same error. This makes handling errors easier. The main place where I'm handling errors (rather than panicking) is in the tests. I had originally written the tests to check the string contents of the errors, but that wasn't possible for the Metadata errors. That lead me to use a base error, which I then figured was better to generalize and use throughout.

Reply:
If that's the case then it's better to use `errors.Wrap`.

Reply:
I'm fine with whatever you want to do here. `fmt.Errorf` also wraps errors by using `%w`. That's how I can use `errors.Is(err, ErrInvalidLabel)` to tell if the returned error is derived from the intended source. I tend to gravitate towards it because it's in the standard library whiles `errors.Wrap` is part of `github.com/pkg/errors`.